runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonLexer.kt

```diff
@@ -0,0 +1,433 @@
+    }
+
+    private fun readKeyword(): JsonToken = when (val ch = peekOrThrow()) {
+        't' -> readLiteral("true", JsonToken.Bool(true))
+        'f' -> readLiteral("false", JsonToken.Bool(false))
+        'n' -> readLiteral("null", JsonToken.Null)
+        else -> fail("Unable to handle keyword starting with '$ch'")
+    }
+
+    private fun readLiteral(expectedString: String, token: JsonToken): JsonToken {
+        consume(expectedString)
+        return token
+    }
+
+    /**
+     * Advance the cursor until next non-whitespace character is encountered
+     * @param peek Flag indicating if the next non-whitespace character should be consumed or peeked
+     */
+    private fun nextNonWhitespace(peek: Boolean = false): Char? {
+        while (peekChar()?.isWhitespace() == true) {
+            idx++
+        }
+        return if (peek) peekChar() else nextOrThrow()
+    }
+
+    /**
+     * Invoke [consume] for each character in [expected]
+     */
+    private fun consume(expected: String) = expected.forEach { consume(it) }
+
+    /**
+     * Assert that the next character is [expected] and advance
+     */
+    private fun consume(expected: Char) {
+        val chr = data[idx].toInt().toChar()
+        check(chr == expected) { "Unexpected char '$chr' expected '$expected'" }
+        idx++
+    }
+
+    /**
+     * Return next byte to consume or null if EOF has been reached
+     */
+    private fun peekByte(): Byte? = data.getOrNull(idx)
+
+    /**
+     * Peek the next character or return null if EOF has been reached
+     *
+     * SAFETY: This assumes ASCII. This is safe because we _only_ use it for tokenization
+     * (e.g. {, }, [, ], <number>, <ws>, etc). When reading object keys or string values [readQuoted] is
+     * used which handles UTF-8. Do not use these single char related functions to directly construct a string!
+     *
+     * NOTE: [readQuoted] uses [decodeToString] which is _MUCH_ faster (~3x) than decoding bytes as
+     * UTF-8 chars one by one on the fly.
+     */
+    private fun peekChar(): Char? = peekByte()?.toInt()?.toChar()
+
+    /**
+     * Peek the next character or throw if EOF has been reached
+     */
+    private fun peekOrThrow(): Char = peekChar() ?: throw IllegalStateException("Unexpected EOF")
+
+    /**
+     * Consume the next character and advance the index or throw if EOF has been reached
+     */
+    private fun nextOrThrow(): Char = peekOrThrow().also { idx++ }
+
+    private fun unexpectedToken(found: Char?, vararg expected: String): Nothing {
+        val pluralModifier = if (expected.size > 1) " one of" else ""
+        val formatted = expected.joinToString(separator = ", ") { "'$it'" }
+        fail("found '$found', expected$pluralModifier $formatted")
+    }
+
+    private fun fail(message: String, offset: Int = idx, cause: Throwable? = null): Nothing {
+        throw DeserializationException("Unexpected JSON token at offset $offset; $message", cause)
+    }
```

Comment:
**Comment**: Raw offsets are not easy to use for debugging when dealing with multiline text. How feasible is it to support line number and position?

Reply:
I gave it some thought but think we would be maybe better off trying to implement better error messages in the deserializer using a path like syntax:

```
Error deserializing path=`.foo.bar[1].blah`; <original cause from lexer>
```

I wasn't sure it was worth the overhead or not to go down this route. I do think in general I agree that the better our error messages are it will help us if we have bugs in the either the lexer/deserializer or in codegen. 

I'm open to ideas.

Reply:
I like the idea of including path information where available but it may not localize the problem very well in some cases. Because the errors represent a failure to parse a well-formed document, the paths cannot localize exactly where the error site is.

In this example:

```json
{
  "foo": "bar",
  "baz": [
    { "key": "val1" },
    { "key": "val2" },
    { "key": "val3" },
    ...
    { "key": "val1023" }
    { "key": "whoops!   ^ I forgot the comma after the last object" },
  ]
}
```

The best path we could derive is `".baz"` right?

In any event, giving a path may help for short documents that can be quickly read by humans. For longer documents that don't fit in an window or require significant mental effort to parse, I still think line:pos is a helpful locator to have alongsize path.

Reply:
Yeah so that's a good point (although I think path _may_ be able to be something like `.baz[n]` where `n` is the closest element). I don't disagree that line:pos may be more helpful but in practice services don't return pretty JSON anyway so there won't be a line number. For a general purpose JSON parser though you are correct that it would probably be more useful. Let me think on it, I definitely want the errors to be useful...